### PR TITLE
[FLINK-31621][table] Add built-in ARRAY_REVERSE function.

### PR DESCRIPTION
**Pull request by liuyongvs**
- What is the purpose of the change
This is an implementation of ARRAY_REVERSE

- Brief change log
ARRAY_REVERSE for Table API and SQL
    
  ```
  array_reverse(array) - Returns an array in reverse order.
  Syntax:
      array_reverse(array)
      
      Arguments:
      array: An ARRAY to be handled.
      
      Returns:
      Returns an array in reverse order.
      If the array itself is null, the function will return null.
   
      > SELECT array_reverse(array(1, 2, 2, NULL));
       NULL, 2, 2, 1
  ```

   See also
    bigquery https://cloud.google.com/bigquery/docs/reference/standard-sql/functions-and-operators#array_reverse

- Verifying this change
This change added tests in CollectionFunctionsITCase

- Does this pull request potentially affect one of the following parts:
Dependencies (does it add or upgrade a dependency): ( no)
The public API, i.e., is any changed class annotated with @Public(Evolving): (yes )
The serializers: (no)
The runtime per-record code paths (performance sensitive): ( no)
Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: ( no)
The S3 file system connector: ( no)
- Documentation
Does this pull request introduce a new feature? (yes)
If yes, how is the feature documented? (docs)

### CODE REVIEW

**Comment by damumu0625** [comment]
Fix the link of ARRAY_REVERSE  in spark and postgresql

**Comment by damumu0625** [comment]
LGTM

**Review by damumu0625 — APPROVED** [review]

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java
Comment (fsk119) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java:63]: Throw better exception details here. It's better we can log the actual input data here for later debug.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java
Comment (fsk119) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java:60]: I think it's not difficult to reverse the array directly here?

**Review by fsk119 — COMMENTED** [review]
Thanks for your contribution.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java
Comment (liuyongvs) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java:63]: @fsk119 other array function does it do. https://github.com/apache/flink/blob/7f47c11da1bbbd6e7650d43742694f2fa0ee8a3f/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayContainsFunction.java#L37

https://github.com/apache/flink/blob/7f47c11da1bbbd6e7650d43742694f2fa0ee8a3f/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayDistinctFunction.java

https://github.com/apache/flink/blob/7f47c11da1bbbd6e7650d43742694f2fa0ee8a3f/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayRemoveFunction.java

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java
Comment (liuyongvs) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java:60]: this tool reverse it quickly and is it common tool.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java
Comment (liuyongvs) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java]: the catch not needs, fixed it

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java
Comment (fsk119) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java]: I still think my idea is better. It does save a lot of time in big data.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java
Comment (snuyanzin) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java:52]: nit: there is no need to create objects for `haystack.size() <= 1`

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java
Comment (snuyanzin) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java:315]: should we have all these 3 lines  in the code?
It seems we cannot differentiate them... So seems not much value

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java
Comment (liuyongvs) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java]: fixed

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java
Comment (liuyongvs) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java]: fixed

**Comment by liuyongvs** [comment]
hi @snuyanzin @fsk119 @damumu0625  fix all your reviews, thanks very much

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java
Comment (fsk119) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java:54]: It's better to use `List<Object>` here to remove the compiler warning.

**Review by fsk119 — COMMENTED** [review]
LGTM with one minor suggestion

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java
Comment (liuyongvs) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java:54]: do i should change other array functions here?

Comment (liuyongvs) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java]: fixed

**Comment by liuyongvs** [comment]
fetch and rebase master and push, waiting for ci passed. thanks @fsk119

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java
Comment (snuyanzin) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java:54]: It looks like we know size of the result array and for `GenericArrayData` we need array.
In this case `List` and `ArrayList` seem to be redundant

**Comment by snuyanzin** [comment]
thanks for addressing comments, in general looks good from my side, i've just added one minor comment

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java
Comment (liuyongvs) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayReverseFunction.java:54]: Great

**Comment by liuyongvs** [comment]
fixed it @snuyanzin

**Comment by snuyanzin** [comment]
lgtm, once ci is passed

**Comment by liuyongvs** [comment]
@snuyanzin ci passed

**Comment by liuyongvs** [comment]
rebase to fix conflicts with array_position which merged just now

**Comment by liuyongvs** [comment]
@snuyanzin @fsk119 will it be merged?

**Pull request merged**
